### PR TITLE
Adding in a function for checking if retries should be performed

### PR DESCRIPTION
**Pull request by smozely**
Function implemented into the config as per #84

### CODE REVIEW

**Comment by cherniavskii** [comment]
I would love to see function support for "retry" option instead!
That way we'll avoid some edge cases.
What do you think @tannerlinsley @smozely ?
It should affect compatibility

**Comment by smozely** [comment]
@Cherniavskii so you mean replace the current ‘retry’ being allowed to be true or a number ... and allow it to also be a object that covers all the config for retries?

I’d be happy to do that, but just want to make sure @tannerlinsley is ok with going that direction

File: src/queryCache.js [hunk 1]
```diff
@@ -313,10 +313,11 @@ export function makeQueryCache() {
 
         // Do we need to retry the request?
         if (
-          // Only retry if the document is visible
-          query.config.retry === true ||
-          query.state.failureCount <= query.config.retry
+          query.config.retryChecker(error, query.state.failureCount) &&
+          (query.config.retry === true ||
+            query.state.failureCount <= query.config.retry)
```
Comment (tannerlinsley) [inline src/queryCache.js]: Reverse this order so that the checker isn't even called if it's not true or less than the retry count?

**Comment by tannerlinsley** [comment]
Also, I'm more fond of the name `shouldRetryFn`. Thoughts?

**Comment by tannerlinsley** [comment]
Yes, I like the idea of just overloading the `retry` option with the ability to be a function, similar to how retry delays work

**Comment by smozely** [comment]
> Yes, I like the idea of just overloading the `retry` option with the ability to be a function, similar to how retry delays work

Ah right, so for that your saying, if I provide a function for retry, then that function should handle everything ... so I'd end up with something like ...

```
config.retry: (failureCount, error) => {
  return failureCount <= 5 && error !== 'NoRetry';
}
```

So I have to implement both the retry count check in my function, and the extra logic ... I'm all cool with that.

**Comment by smozely** [comment]
I've updated with the overloading of the `retry` config item

**Comment by cherniavskii** [comment]
> So I have to implement both the retry count check in my function, and the extra logic

Yeah, I think it's still way better than dealing with edge cases like:
```js
{
  retry: true,
  retryChecker: () => false,
  // what should happen?
}
```

File: src/queryCache.js [hunk 2]
```diff
@@ -313,10 +313,12 @@ export function makeQueryCache() {
 
         // Do we need to retry the request?
         if (
-          // Only retry if the document is visible
           query.config.retry === true ||
-          query.state.failureCount <= query.config.retry
+          (typeof query.config.retry == 'number' &&
```
Comment (cherniavskii) [inline src/queryCache.js]: I don't think we need this check here. If retry is a function, this condition will be falsy

File: src/queryCache.js [hunk 3]
```diff
@@ -313,10 +313,12 @@ export function makeQueryCache() {
...
+            query.state.failureCount <= query.config.retry) ||
+          query.config.retry(query.state.failureCount, error)
```
Comment (cherniavskii) [inline src/queryCache.js]: We need if retry is a function before calling it

File: src/tests/useQuery.test.js [hunk 4]
```diff
@@ -249,6 +249,37 @@ describe('useQuery', () => {
...
+    const queryFn = jest.fn()
+
+    queryFn.mockImplementation(() => {
+      return Promise.reject('NoRetry')
+    })
+
+    function Page() {
+      const { status, failureCount } = useQuery('test', queryFn, {
+        retryDelay: 1,
+        retry: (failureCount, error) => error !== 'NoRetry',
```
Comment (cherniavskii) [inline src/tests/useQuery.test.js]: This test will pass even without this feature implemented:
`retry` is a function, therefore `retry === true` is `false` and `query.state.failureCount <= query.config.retry` is false as well, so query won't be refetched
https://github.com/tannerlinsley/react-query/blob/2b6a711cf38acaa891bf0db3a6d871247abe9b9a/src/queryCache.js#L315-L319

My suggestion is to allow retry once and expect query to fail 2 times

**Comment by smozely** [comment]
@cherniavskii Thanks for pointing out the mistake with those typeof checks, I also expanded another test which was passing for the wrong reasons with my change

**Comment by cherniavskii** [comment]
Awesome!

**Review by cherniavskii — APPROVED** [review]
LGTM!

**Comment by smozely** [comment]
Whoop! thanks folks ... now to replace my home-grown solution with this!

**Pull request merged**